fuzz-test-suite/fuzzamericanoption.cpp

```diff
@@ -0,0 +1,132 @@
+ Copyright (C) 2005 Joseph Wang
+
+ This file is part of QuantLib, a free-software/open-source library
+ for financial quantitative analysts and developers - http://quantlib.org/
+
+ QuantLib is free software: you can redistribute it and/or modify it
+ under the terms of the QuantLib license.  You should have received a
+ copy of the license along with this program; if not, please email
+ <quantlib-dev@lists.sf.net>. The license is also available online at
+ <http://quantlib.org/license.shtml>.
+
+ This program is distributed in the hope that it will be useful, but WITHOUT
+ ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ FOR A PARTICULAR PURPOSE.  See the license for more details.
+*/
+
+#include <ql/any.hpp>
+#include <ql/exercise.hpp>
+#include <ql/termstructures/volatility/equityfx/blackconstantvol.hpp>
+#include <ql/instruments/vanillaoption.hpp>
+#include <ql/math/distributions/normaldistribution.hpp>
+#include <ql/math/functional.hpp>
+#include <ql/math/integrals/gausslobattointegral.hpp>
+#include <ql/math/integrals/integral.hpp>
+#include <ql/math/randomnumbers/rngtraits.hpp>
+#include <ql/math/statistics/incrementalstatistics.hpp>
+#include <ql/pricingengines/vanilla/analyticeuropeanengine.hpp>
+#include <ql/pricingengines/vanilla/baroneadesiwhaleyengine.hpp>
+#include <ql/pricingengines/vanilla/bjerksundstenslandengine.hpp>
+#include <ql/pricingengines/vanilla/fdblackscholesshoutengine.hpp>
+#include <ql/pricingengines/vanilla/fdblackscholesvanillaengine.hpp>
+#include <ql/pricingengines/vanilla/juquadraticengine.hpp>
+#include <ql/pricingengines/vanilla/qdfpamericanengine.hpp>
+#include <ql/pricingengines/vanilla/qdplusamericanengine.hpp>
+#include <ql/termstructures/yield/flatforward.hpp>
+#include <ql/time/daycounters/actual360.hpp>
+#include <ql/time/daycounters/thirty360.hpp>
+#include <ql/time/calendars/nullcalendar.hpp>
+#include <ql/utilities/dataformatters.hpp>
+#include <fuzzer/FuzzedDataProvider.h>
+#include <limits>
+#include <map>
+
+
+using namespace QuantLib;
+
+namespace {
+
+    struct AmericanOptionData {
+        Option::Type type;
+        Real strike;
+        Real s;       // spot
+        Rate q;       // dividend
+        Rate r;       // risk-free rate
+        Time t;       // time to maturity
+        Volatility v; // volatility
+    };
+
+}
+
+AmericanOptionData fuzzedAmericanOptionData(FuzzedDataProvider& fdp) {
+    return AmericanOptionData{
+        .type = fdp.PickValueInArray({Option::Type::Put, Option::Type::Call}),
+        .strike = fdp.ConsumeFloatingPoint<Real>(),
+        .s = fdp.ConsumeFloatingPointInRange<Real>(0.0, std::numeric_limits<Real>::max()),
+        .q = fdp.ConsumeProbability<Rate>(),
+        .r = fdp.ConsumeProbability<Rate>(),
+        .t = fdp.ConsumeFloatingPointInRange<Time>(0.0, std::numeric_limits<Real>::max()),
+        // TODO(silvergasp): Confirm this should be between 0.0-1.0
+        .v = fdp.ConsumeProbability<Volatility>(),
```

Comment:
I wasn't sure about this metric how is is volatility normally characterized? i.e. as a value between 0.0-1.0 or is this any positive number?

Reply:
Ok so it's seems like the algorithm would be valid for any positive number. But in the real world this number would rarely if ever reach 100.0. Would that be a reasonable assesment? Unless you have a specific value in mind I'll cap this at 10.0 and we can see how that tracks.

Reply:
Ok for me.